semantic/constraint.go

```diff
@@ -0,0 +1,334 @@
+		// -> typeof(f) = (typeof(a), typeof(b)) => typeof(f(a:1, b:2))
+		operator := c.tenv[n.Callee]
+		operand := make(map[string]Substitutable, len(n.Arguments.Properties))
+		for i, prop := range n.Arguments.Properties {
+			operand[prop.Key.Name] = c.tenv[n.Arguments.Properties[i].Value]
+		}
+		funcType := funcTypeExpression{
+			params:     operand,
+			returnType: tv,
+		}
+		c.cons = append(c.cons, Constraint{
+			left:  operator,
+			right: funcType,
+		})
+	case *UnaryExpression:
+		// Inference Rule: Unary Expression
+		// --------------------------------
+		// x = (op) expression
+		//
+		// -> typeof(x) = bool
+		// -> typeof(expression) = bool
+		//
+		// Unary expressions yield boolean types
+		c.cons = append(c.cons, Constraint{
+			left:  tv,
+			right: Bool,
+		})
+		switch n.Operator {
+		// Unary operators must act on boolean types
+		case ast.NotOperator:
+			c.cons = append(c.cons, Constraint{
+				left:  c.tenv[n.Argument],
+				right: Bool,
+			})
+		}
+	case *LogicalExpression:
+		// Inference Rule: Logical Expression
+		// ----------------------------------
+		// x = left (op) right
+		//
+		// -> typeof(x) = bool
+		// -> typeof(left) = bool
+		// -> typeof(right) = bool
+		//
+		// Logical expressions yield boolean types
+		// Logical operators act on boolean types
+		c.cons = append(c.cons,
+			Constraint{
+				left:  tv,
+				right: Bool,
+			},
+			Constraint{
+				left:  c.tenv[n.Left],
+				right: Bool,
+			},
+			Constraint{
+				left:  c.tenv[n.Right],
+				right: Bool,
+			},
+		)
+	case *BinaryExpression:
+		switch n.Operator {
+		// Inference Rule: Arithmetic Operators
+		// ------------------------------------
+		// x = a ( + , - , * , / ) b
+		//
+		// -> typeof(x) = typeof(a) = typeof(b)
+		//
+		// TODO: Only Int, UInt, Float, and String types supported
+		case ast.AdditionOperator:
+			c.cons = append(c.cons,
+				Constraint{
+					left:  tv,
+					right: c.tenv[n.Left],
+				},
+				Constraint{
+					left:  tv,
+					right: c.tenv[n.Right],
+				},
+				Constraint{
+					left:  c.tenv[n.Right],
+					right: c.tenv[n.Left],
+				},
+			)
+		case ast.SubtractionOperator:
+			c.cons = append(c.cons,
+				Constraint{
+					left:  tv,
+					right: c.tenv[n.Left],
+				},
+				Constraint{
+					left:  tv,
+					right: c.tenv[n.Right],
+				},
+				Constraint{
+					left:  c.tenv[n.Right],
+					right: c.tenv[n.Left],
+				},
+			)
+		case ast.MultiplicationOperator:
+			c.cons = append(c.cons,
+				Constraint{
+					left:  tv,
+					right: c.tenv[n.Left],
+				},
+				Constraint{
+					left:  tv,
+					right: c.tenv[n.Right],
+				},
+				Constraint{
+					left:  c.tenv[n.Right],
+					right: c.tenv[n.Left],
+				},
+			)
+		case ast.DivisionOperator:
+			c.cons = append(c.cons,
+				Constraint{
+					left:  tv,
+					right: c.tenv[n.Left],
+				},
+				Constraint{
+					left:  tv,
+					right: c.tenv[n.Right],
+				},
+				Constraint{
+					left:  c.tenv[n.Right],
+					right: c.tenv[n.Left],
+				},
+			)
```

Comment:
All the above cases look the same. Is that right? If so can you collapse them into a single case? You can comma separate the cases like so:

`case ast.MultiplicationOperator,ast.DivisionOperator,...:`

Reply:
That is right. I didn't know you could do that.